content/match-types-spec.md

```diff
@@ -0,0 +1,354 @@
+  but the author already accepted a change with a workaround at
+  https://github.com/Katrix/perspective/pull/1
+* 1 library has 3 occurrences of using an abstract type constructor too "concretely":
+  https://github.com/kory33/s2mc-test/blob/d27c6e85ad292f8a96d7d51af7ddc87518915149/protocol-core/src/main/scala/io/github/kory33/s2mctest/core/generic/compiletime/Tuple.scala#L16
+  defined at https://github.com/kory33/s2mc-test/blob/d27c6e85ad292f8a96d7d51af7ddc87518915149/protocol-core/src/main/scala/io/github/kory33/s2mctest/core/generic/compiletime/Generic.scala#L12
+  It could be replaced by a concrete `class Lock[A](phantom: A)` instead.
+
+The only case for which there exists no workaround that we know of is the extractor for `scala.Enumeration`-based `Value` classes.
+
+### Other concerns
+
+Ideally, this proposal would be first implemented as *warnings* about illegal cases, and only later made errors.
+Unfortunately, the presence of the abstract type constructor case makes that impossible.
+Indeed, because of it, a pattern that is legal at definition site may become illegal after some later substitution.
+
+Consider for example the standard library's very own `Tuple.InverseMap`:
+
+```scala
+/** Converts a tuple `(F[T1], ..., F[Tn])` to `(T1,  ... Tn)` */
+type InverseMap[X <: Tuple, F[_]] <: Tuple = X match {
+  case F[x] *: t => x *: InverseMap[t, F]
+  case EmptyTuple => EmptyTuple
+}
+```
+
+If we instantiate `InverseMap` with a class type parameter, such as `InverseMap[X, List]`, the first case gets instantiated to
+```scala
+case List[x] *: t => x *: InverseMap[t, List]
+```
+which is legal.
+
+However, nothing prevents us a priori to instantiate `InverseMap` with an illegal type constructor, for example
+```scala
+type IsSeq[t <: Seq[Any]] = t
+InverseMap[X, IsSeq]
+```
+which gives
+```scala
+case IsSeq[x] *: t => x *: InverseMap[t, IsSeq]
+```
+
+These instantiatiations happen deep inside the type checker, during type computations.
+Since types are cached, shared and reused in several parts of the program, by construction, we do not have any source code position information at that point.
+That means that we cannot report *warnings*.
+
+We can in fact report *errors* by reducing to a so-called `ErrorType`, which is aggressively propagated.
+This is what we do in the proposed implementation (unless using `-source:3.3`).
+
+### Open questions
+
+None at this point.
+
+## Alternatives
+
+The specification is more complicated than we initially wanted.
+At the beginning, we were hoping that we could restrict match cases to class type constructors only.
+The quantitative study however revealed that we had to introduce support for abstract type constructors and for type member extractors.
+
+As already mentioned, the standard library itself contains an occurrence of an abstract type constructor in a pattern.
+Making that an error would mean declaring the standard library itself bankrupt, which was not a viable option.
```

Comment:
“bankrupt” is a colorful metaphor here, but I'm left wondering what underlying literal meaning is intended. like, we'd need to make an incompatible change to the standard library that would break some code? (how much code? compile-time breakage, runtime breakage?) is making such a change impossible? highly undesirable? at least somewhat undesirable?